fw/http_parser.c

```diff
@@ -3558,6 +3584,449 @@ __req_parse_x_forwarded_for(TfwHttpMsg *hm, unsigned char *data, size_t len)
+			__FSM_I_MOVE_fixup(Req_I_Fwd_For_Sep_End, 1, 0);
+
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_Start) {
+		if (unlikely(IS_CRLFWS(c)))
+			return CSTR_NEQ;
+		if (likely(c == '"'))
+			__FSM_I_MOVE_fixup(Req_I_Fwd_Host_Start_Quoted, 1, 0);
+		/* Fall through */
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_Unquoted) {
+		/* See Req_UriAuthority processing. */
+		__fsm_sz = 0;
+
+		while (likely(isalnum(c) || c == '.' || c == '-')) {
+			__fsm_sz++;
+			c = *(p + __fsm_sz);
+			if (unlikely(__data_off(p + __fsm_sz) >= len))
+				__FSM_I_POSTPONE_fixup(Req_I_Fwd_Host_Unquoted,
+						       __fsm_sz,
+						       TFW_STR_VALUE);
+		}
+		__FSM_I_MOVE_fixup(Req_I_Fwd_Host_End_Unquoted,
+				   __fsm_sz,
+				   TFW_STR_VALUE);
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_Start_Quoted) {
+		if (likely(c == '['))
+			__FSM_I_MOVE_fixup(Req_I_Fwd_Host_v6_Quoted_Start,
+					   1, TFW_STR_VALUE);
+		/* Block empty quotes */
+		if (unlikely(c == '"'))
+			return CSTR_NEQ;
+		/* Fall through */
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_Quoted) {
+		/* See Req_UriAuthority processing. */
+		__fsm_sz = 0;
+
+		while (likely(isalnum(c) || c == '.' || c == '-')) {
+			__fsm_sz++;
+			c = *(p + __fsm_sz);
+			if (unlikely(__data_off(p + __fsm_sz) >= len))
+				__FSM_I_POSTPONE_fixup(Req_I_Fwd_Host_Quoted,
+						       __fsm_sz,
+						       TFW_STR_VALUE);
+		}
+		__FSM_I_MOVE_fixup(Req_I_Fwd_Host_End_Quoted,
+				   __fsm_sz,
+				   TFW_STR_VALUE);
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_v6_Quoted_Start) {
+		/* Block empty braces */
+		if (unlikely(c == ']'))
+			return CSTR_NEQ;
+		/* Fall through */
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_v6_Quoted) {
+		/* See Req_UriAuthorityIPv6 processing. */
+		__fsm_sz = 0;
+
+		while (likely(isxdigit(c) || c == ':')) {
+			__fsm_sz++;
+			c = *(p + __fsm_sz);
+			if (unlikely(__data_off(p + __fsm_sz) >= len))
+				__FSM_I_POSTPONE_fixup(Req_I_Fwd_Host_v6_Quoted,
+						       __fsm_sz,
+						       TFW_STR_VALUE);
+		}
+		if (likely(c == ']'))
+			__FSM_I_MOVE_fixup(Req_I_Fwd_Host_End_Quoted,
+					   __fsm_sz + 1, TFW_STR_VALUE);
+
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_End_Quoted) {
+		if (c == ':')
+			__FSM_I_MOVE_fixup(Req_I_Fwd_Host_Port_Quoted,
+					   1, 0);
+		__FSM_I_JMP(Req_I_Fwd_Next_Or_Finish_Quoted);
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_End_Unquoted) {
+		if (c == ':')
+			__FSM_I_MOVE_fixup(Req_I_Fwd_Host_Port_Unquoted,
+					   1, 0);
+		__FSM_I_JMP(Req_I_Fwd_Next_Or_Finish);
+	}
+
+	__FSM_STATE(Req_I_Fwd_Host_Port_Unquoted) {
+		__fsm_sz = __data_remain(p);
+		__fsm_n = __parse_ulong_ws_delim(p, __fsm_sz,
+						 (unsigned long*)&parser->port,
+						 USHRT_MAX);
+		switch (__fsm_n) {
+		case CSTR_BADLEN:
+		case CSTR_NEQ:
+			return CSTR_NEQ;
+		case CSTR_POSTPONE:
+			if (parser->port == 0 || parser->port > 65535)
+				return CSTR_NEQ;
+			__FSM_I_MOVE_fixup(Req_I_Fwd_Host_Port_Unquoted,
+					   __fsm_sz, TFW_STR_VALUE);
+		default:
+			if (parser->port == 0 || parser->port > 65535)
```

Comment:
I believe `parser->port` must be `unsigned int`. __builtin_uaddl_overflow()` in `__parse_ulong()` might write larger than 2 bytes to `parser->port` before `if` checks the limit. At the moment, thank to data alignment, we have extra 2 bytes in the union, but this might change and we'll get a vulnerability.

BTW with the current code `parser->port` is unsugned short, so it's never > 65535.

The same for HTTP/2